release-22.1: changefeedccl: add non-batching retries to kafka sink

samiskin opened the pull request:
Backport 1/1 commits from https://github.com/cockroachdb/cockroach/pull/86138
Backport 1/1 commits from https://github.com/cockroachdb/cockroach/pull/88983
Backport 1/1 commits from https://github.com/cockroachdb/cockroach/pull/89173

/cc @cockroachdb/release 

---

A recurring painpoint of changefeeds is the "message too large" kafka
issue where a fixed size batch may sometimes end up being too many bytes
in size due to varying message sizes, but there was no mechanism to
reduce the batch size dynamically, so the changefeed would simply fail.

This PR adds an internal retry mechanism to the kafka sink, where upon
observing a messagetoolarge error the sink will pause all new incoming
messages, attempt to retry those messages with a progressively smaller
batch size, then once one is found resumes operation with the original
configuration.

Release note (bug fix): Changefeeds emitting to Kafka upon receiving a
"message too large" error will now halve the size of their batches until
it either succeeds or a batch size of 1 fails.

Release justification: high priority bug fix, new code is mostly only ran in
cases where an unrecoverable error would've occurred previously and can
be toggled with a cluster setting.

---

Release justification: high priority bug fix

samiskin committed:
changefeedccl: add non-batching retries to kafka sink

A recurring painpoint of changefeeds is the "message too large" kafka
issue where a fixed size batch may sometimes end up being too many bytes
in size due to varying message sizes, but there was no mechanism to
reduce the batch size dynamically, so the changefeed would simply fail.

This PR adds an internal retry mechanism to the kafka sink, where upon
observing a messagetoolarge error the sink will pause all new incoming
messages, attempt to retry those messages with a progressively smaller
batch size, then once one is found resumes operation with the original
configuration.

Release note (bug fix): Changefeeds emitting to Kafka upon receiving a
"message too large" error will now halve the size of their batches until
it either succeeds or a batch size of 1 fails.

Release justification: high priority bug fix, new code is mostly only ran in
cases where an unrecoverable error would've occurred previously and can
be toggled with a cluster setting.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87048)
<!-- Reviewable:end -->

samiskin committed:
changefeedccl: avoid extra internal retry for buffered emits

With the new parallel event consumer changes even when the sink errors,
storing a value in flushErr, there can be many more emits with no flush,
which on this test would result in constant internal retries occuring.
This change just makes internal retries only occur if no flushErr has
occured, since we've already errored and all we want is to let
everything flush out.  This fixes a flake in
TestChangefeedKafkaMessageTooLarge.

Release note: None

samiskin committed:
changefeedccl: fix shutdown deadlock on kafka internal retry

This fixes a test flake where if the changefeed was cancelled right
after registering an inflight message to be sent, the message wouldn't
send but the kafka sink worker would remain blocked expecting one to
arrive. stopWorkerCh was never closed since kafkaSink is only closed
after eventConsumer is closed, but eventConsumer was blocked on sending
a message.

Release note (bug fix): changefeeds no longer deadlock when cancelling
during an internal kafka sink retry.

samiskin commented:
The only difference between this and master is that the option defaults to false + the tests are each ran in their own `kafkaTest(...)` call as for some reason it would fail when reusing the same sink even though this error didn't happen on master.  I was stuck for a while on seeing `TestChangefeedFlushesSinkToReleaseMemory` fail, but turns out thats a flake that happens even with no changes as shown by this no-op test branch CI failure https://github.com/cockroachdb/cockroach/pull/89943

miretskiy reviewed (changes requested):
I don't know if this batch retrying works at all.

miretskiy commented:
Informs https://github.com/cockroachdb/cockroach/issues/90029

shermanCRL commented:
Is a backport to v22.1 plausible or advisable? There are large users who are on that version and are likely to be for a while.
